Fix #4

simoneduca opened the pull request:
@rogerhutchings note that I left the client in because it handles the hosts depending on the environment for me. I don't think we've got concrete evidence that making the call directly would be faster, so I'd like to leave like this, if you agree.

Commit:
Fix #4

rogerhutchings commented:
I'm not sure it'd be faster, but it does remove an extra dependency for your code (so if we break the client it won't break this), and that is worth considering.

The environment handling should be pretty simple:

```
const API_URL = (process.env.NODE_ENV === 'production') ? url_1 : url_2;
```

You can then just use `fetch`/whatever to do your request.

Commit:
Include review comments. Remove panoptes client.

simoneduca commented:
@rogerhutchings thanks, that should be it.

Commit:
Better comments and more sensible naming.

camallen commented on app.js:
```diff
@@ -66,10 +66,10 @@ app.use(function(req, res, next) {
-  if (isValidToken(req.query.token)) {
-    sessionExists(req.query.token).then(function() {
+  var apiToken = req.query.token;
+  if (isValidToken(apiToken)) {
```
out of interest, does the `isValidToken` function return a boolean value for use in the conditional?

simoneduca replied:
Not sure I understand what you mean by `for use`; It does return `true` if the token is valid, `false` if it isn't.

camallen replied:
my bad, i was looking at a single commit and not the whole set of changes...as you were.

camallen commented on app.js:
```diff
@@ -26,10 +66,17 @@ app.use(function(req, res, next) {
+      res.json({ token });
+    })
+  } else {
+    console.log('Invalid token')
```
You should probably respond with a HTTP 422 or something here that is useful for the client making the call to understand why it didn't work.

simoneduca replied:
Aha! Good one, thanks.

camallen replied:
e.g. [here](https://github.com/zooniverse/static-crop/blob/f01cca3727aad253c16ae3ea95f75337e3c64931/server.js#L105) and other places in that link we're providing some info about what went wrong in the req if error and if the params are invalid, etc.

simoneduca replied:
Included your comment. How does it look now?

Commit:
Send invalid data response.

simoneduca commented:
@camallen included your comment. How does it look now?

camallen commented on app.js:
```diff
@@ -26,10 +66,17 @@ app.use(function(req, res, next) {
-  res.json({ token });
+  var apiToken = req.query.token;
+  if (isValidToken(apiToken)) {
+    getApiUser(apiToken).then(function() {
```
what happens when this function throws an error?

simoneduca replied:
It should tell the client.

simoneduca replied:
I'll put that in.

camallen replied:
How? via the `.then` function that responds with a json token?

rogerhutchings replied:
I don't think it responds with anything - there should be a `catch` here for when `getApiUser` throws an error that sends a response

simoneduca replied:
Rog beat me to it. Sending the client an error code in the catch block

rogerhutchings commented on package.json:
```diff
@@ -11,6 +11,5 @@
   "dependencies": {
     "express": "^4.14.0",
     "firebase": "^3.3.0",
```
trailing comma here

camallen commented on app.js:
```diff
@@ -26,10 +66,17 @@ app.use(function(req, res, next) {
+  var apiToken = req.query.token;
+  if (isValidToken(apiToken)) {
+    getApiUser(apiToken).then(function() {
+      const token = firebase.auth().createCustomToken(req.query.uid);
```
why does the req contain the UID of the user? I thought you'd use the response form the /me route to get the user's login (unique in panoptes).

simoneduca replied:
I am using the user's login when making the request from Vox

simoneduca replied:
But it probably makes more sense not to pass the user's login with the request and to use it here instead. Is that what you mean?

camallen replied:
Could i impersonate a user this way if i modify the req params? Remember the bearer token is per user on the api, so it's all you need to ID a user.

simoneduca replied:
I see. So how about something like this instead `firebase.auth().createCustomToken(user.login);`? Btw, Firebase method doesn't like receiving a token, otherwise I would have just passed that to it.

camallen replied:
you need to pass firebase a UID (user.login) and you can include other details that you want the firebase token to employ too from mem.

camallen replied:
and yeah if you have the api user then you can use the user.login string as the UID here.

simoneduca replied:
Correct https://firebase.google.com/docs/auth/server/create-custom-tokens but I didn't think we needed anything else other than the uid.

Commit:
Include some review comments

rogerhutchings committed:
Update package.json

- "main" isn't needed when a project is standalone
- fix licence code
- add urls etc
- don't publish on npm

Commit:
Use current Express syntax. Previous is deprecated.

Commit:
Merge branch 'fix-4' of github.com:zooniverse/firebase-token-generator into fix-4

simoneduca commented:
Didn't notice your commit @rogerhutchings, sorry. The license was fixed in a previous PR.

simoneduca commented:
May I suggest that if people have more questions/issues to open them as new issues? It'd be easier to keep track and we can merge this.

camallen commented:
LGTM, if we're commenting on the code in the PR we should not move them to another issue as it's relevant to this code review.

simoneduca commented:
Fair enough; mind merging @camallen ?

camallen commented:
I've stated it looks good, go ahead at your leisure if you're happy.
